### PR TITLE
Add sampling handler and sampling strategy store

### PR DESCRIPTION
**Pull request by black-adder**
This adds a tchannel sampling handler and a static sampling strategy store that reads sampling strategies from a json file.

I added a factory for this so that when we move adaptive sampling over to open source, we can easily extend the existing factory with adaptive sampling and all its flags (I've counted 15 flags at the time of writing).

Locations of files might be up for debate (should this belong in pkg and plugin?)
Do I even need a separate sampling strategy store? should I just use the sampling handler interface and have a factory for that?

Fixes https://github.com/jaegertracing/jaeger/issues/661

Signed-off-by: Won Jun Jang <wjang@uber.com>

### CODE REVIEW

File: plugin/sampling/strategystore/static/options.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+package static
+
+import (
+	"flag"
+
+	"github.com/spf13/viper"
+)
+
+const (
+	samplingStrategies = "sampling.strategies"
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/static/options.go]: I would prefer a bit more explicit name "sampling.strategies-file"

File: plugin/sampling/strategystore/static/strategy_store.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+// GetSamplingStrategy implements StrategyStore#GetSamplingStrategy.
+func (h *strategyStore) GetSamplingStrategy(serviceName string) (*sampling.SamplingStrategyResponse, error) {
+	if strategy, ok := h.serviceStrategies[serviceName]; ok {
+		return strategy, nil
+	}
+	return h.defaultStrategy, nil
+}
+
+// TODO good candidate for a global util function
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/static/strategy_store.go]: how much overlap is in this code with what the client is doing when parsing JSON response from the agent? Maybe we can combine that all into jaeger-lib?

**Comment by pavolloffay** [comment]
@black-adder does this fix https://github.com/jaegertracing/jaeger/issues/661?

File: plugin/sampling/strategystore/static/strategy_store.go (hunk 2 above)
Comment (black-adder) [reply plugin/sampling/strategystore/static/strategy_store.go]: This TODO was originally meant for creating a util function for reading a json file and unmarshalling it. I didn't have as lofty goal of moving the client and this logic into jaeger-lib. I don't think it's similar enough to warrant that.

File: pkg/sampling/strategystore/constants.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) 2018 The Jaeger Authors.
```
Comment (yurishkuro) [inline pkg/sampling/strategystore/constants.go]: yes, the `pkg` is rather suspicious for this. It is meant for things not directly related to Jaeger. The simplest approach would be to define the storage interface in `cmd/collector/sampling/strategystore/interface.go`

File: pkg/sampling/strategystore/factory.go [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	"github.com/uber/jaeger-lib/metrics"
+	"go.uber.org/zap"
+)
+
+// Factory defines an interface for a factory that can create implementations of different strategy storage components.
+// Implementations are also encouraged to implement plugin.Configurable interface.
+//
+// See also
+//
+// plugin.Configurable
```
Comment (yurishkuro) [inline pkg/sampling/strategystore/factory.go]: Configurable also has AddFlags(flagSet *flag.FlagSet). Do you actually need to require Configurable methods in the Factory interface?

File: pkg/sampling/strategystore/constants.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package strategystore
+
+import (
+	"github.com/jaegertracing/jaeger/thrift-gen/sampling"
+)
+
+const (
```
Comment (yurishkuro) [inline pkg/sampling/strategystore/constants.go]: is this needed in the "interface" package? Could it not be placed in the static impl?

File: pkg/sampling/strategystore/strategy.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package strategystore
+
+// Strategy defines a sampling strategy. Type can be "probabilistic" or "ratelimiting"
+// and Param will represent "sampling probability" and "max traces per second" respectively.
+type Strategy struct {
```
Comment (yurishkuro) [inline pkg/sampling/strategystore/strategy.go]: these types look like private intermediate model used by static impl. Or did you mean the StrategyStorage interface to return these?

File: plugin/sampling/strategystore/factory_config.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+// FactoryConfig tells the Factory what sampling type it needs to create.
+type FactoryConfig struct {
+	StrategyStoreType string
+}
+
+// FactoryConfigFromEnv reads the desired sampling type from the SAMPLING_TYPE environment variable. Allowed values:
+//   * `static` - built-in
+//   * `adaptive` - built-in // TODO
+func FactoryConfigFromEnv() FactoryConfig {
```
Comment (yurishkuro) [inline plugin/sampling/strategystore/factory_config.go]: are you planning to add changes to collector.main() here?   Right now this function is not used.

File: pkg/sampling/strategystore/factory.go (hunk 4 above)
Comment (black-adder) [reply pkg/sampling/strategystore/factory.go]: It doesn't have to be but it makes life easier.

File: pkg/sampling/strategystore/strategy.go (hunk 6 above)
Comment (black-adder) [reply pkg/sampling/strategystore/strategy.go]: made private

File: pkg/sampling/strategystore/constants.go (hunk 3 above)
Comment (black-adder) [reply pkg/sampling/strategystore/constants.go]: done, should I have the static/adaptive implementations in plugin or should I move the whole thing over to cmd/collector/app?

**Pull request merged**